algorithms.tex

```diff
@@ -1412,6 +1412,15 @@
 behave as if they are implemented by calling \tcode{begin} and \tcode{end} on
 the \tcode{Range} and dispatching to the overload that takes separate
 iterator and sentinel arguments.
+
+\pnum
+\newtxt{Some algorithms declare both an overload that takes a \tcode{Range}
+and an \tcode{Iterator}, and an overload that takes two \tcode{Range}
+arguments. Since an array type~(\cxxref{basic.compound}) both satisifies
+\tcode{Range} and decays to a pointer~(\cxxref{conv.array}) which satisfies
+\tcode{Iterator}, such overloads are ambiguous when an array is passed as
+the second argument. Implementations shall provide a mechanism to resolve
+this ambiguity in favor of the overload that takes two ranges.}
```

Comment:
> Whether an iterator or range is passed by value or by forwarding reference is technically observable

I've been operating under the assumption that given unordered template parameters it's unobservable whether
function parameters are taken by value or forwarding reference. That's only true, however, if the initialization of copies from the arguments is sequenced before any invocation of user code. That is clearly a ridiculous standard to require of implementations; certainly mine doesn't meet it. Every one of my range algorithms, for example, calls `begin` and `end` on the range and perfect-forwards the other arguments into an iterator algorithm where they will eventually be copied/moved. I would rather have fewer total copies/moves than have the first set of copies/moves occur in the proper sequence.

I wonder if we'd be better served by raising the level of semantic abstraction for the algorithm specifications as we did for customization points. Instead of specifying the existence and behavior of two overloaded function templates named, e.g., `find_if`, we would specify the effects of the expression `ranges::find_if(i, s, t)` and `ranges::find_if(r, t)` when the types of named parameters satisfy the appropriate concepts. Template parameters and any ordering thereof never enter into the picture. That style specification would even allow - but not require - implementation of `find_if` as an object instead of a set of function template overloads.